tokenclient/tokenFlows.go

```diff
@@ -0,0 +1,196 @@
+	Context context.Context
+	// Request parameters that shall be overwritten or added to the payload
+	Params map[string]string
+}
+
+type TokenFlows struct {
+	identity *env.Identity
+	options  Options
+	tokenURI string
+}
+
+type ClientError struct {
+	msg string
+	err error
+}
+
+func (e *ClientError) Error() string {
+	if e.err == nil {
+		return e.msg
+	}
+	return fmt.Sprintf("%s: %s", e.msg, e.err.Error())
+}
+
+type tokenResponse struct {
+	Token string `json:"access_token"`
+}
+
+const (
+	tokenEndpoint              string = "/oauth2/token" //nolint:gosec
+	grantTypeParameter         string = "grant_type"
+	grantTypeClientCredentials string = "client_credentials"
+	clientIDParameter          string = "client_id"
+	clientSecretParameter      string = "client_secret"
+)
+
+// NewTokenFlows initializes token flows
+//
+// identity provides credentials and url to authenticate client with identity service
+// options specifies rest client and its tls config, both can be overwritten.
+// Note: Setup of default tls config is not supported for windows os. Module crypto/x509 supports SystemCertPool with go 1.18 (https://go-review.googlesource.com/c/go/+/353589/)
+func NewTokenFlows(identity *env.Identity, options Options) (*TokenFlows, *ClientError) {
+	t := new(TokenFlows)
+	t.identity = identity
+	if options.HTTPClient == nil {
+		if options.TLSConfig == nil && identity.IsCertificateBased() {
+			defaultConfig, err := defaultTLSConfig(identity)
+			if err != nil {
+				return nil, err
+			}
+			options.TLSConfig = defaultConfig
+		}
+		options.HTTPClient = defaultHTTPClient(options.TLSConfig)
+	}
+	t.options = options
+	t.tokenURI = identity.GetURL() + tokenEndpoint
+	return t, nil
+}
+
+// ClientCredentials implements the client credentials flow (RFC 6749, section 4.4).
+// Clients obtain an access token outside of the context of a user.
+// It is used for non interactive applications (a CLI, a batch job, or for service-2-service communication) where the token is issued to the application itself,
+// instead of an end user for accessing resources without principal propagation.
+//
+// options allows to provide a request context and optionally additional request parameters
+func (t *TokenFlows) ClientCredentials(customerTenantHost string, options RequestOptions) (auth.Token, *ClientError) {
+	data := url.Values{}
+	data.Set(grantTypeParameter, grantTypeClientCredentials)
+	data.Set(clientIDParameter, t.identity.GetClientID())
+	if t.identity.GetClientSecret() != "" {
+		data.Set(clientSecretParameter, t.identity.GetClientSecret())
+	}
+	for name, value := range options.Params {
+		data.Set(name, value) // potentially overwrites data which was set before
+	}
+	targetURL, err := t.getURL(customerTenantHost)
+	if err != nil {
+		return nil, err
+	}
+	ctx := options.Context
+	if ctx == nil {
+		log.Printf("uses context.TODO as fallback, as no context is provided with RequestOptions")
+		ctx = context.TODO()
+	}
```

Comment:
Perhaps we could make the consumer choose a context by having `ctx` as the first argument.

Reply:
@jkbschmid you mean:
instead of
`ClientCredentials(customerTenantHost string, options RequestOptions)`
design it like:
`ClientCredentials(ctx context.Context, customerTenantHost string, options RequestOptions)`
?

Reply:
done

Reply:
Yes, exactly. That's how it's recommended (see https://pkg.go.dev/context#pkg-overview).